File: services/repository-ingestion/src/utils/embedding-service.ts
import { logger } from './logger';
import axios from 'axios';

export interface EmbeddingOptions {
  model?: string;
  dimensions?: number;
  normalize?: boolean;
}

export interface EmbeddingResult {
  embedding: number[];
  dimensions: number;
  model: string;
  tokens: number;
}

export class EmbeddingService {
  private ollamaUrl: string;
  private defaultModel: string;
  private cache: Map<string, EmbeddingResult> = new Map();

  constructor() {
    this.ollamaUrl = process.env.OLLAMA_URL || 'http://localhost:11434';
    this.defaultModel = process.env.OLLAMA_MODEL || 'nomic-embed-text';
  }

  async generateEmbedding(
    text: string, 
    options: EmbeddingOptions = {}
  ): Promise<number[]> {
    try {
      // Check cache first
      const cacheKey = this.getCacheKey(text, options);
      const cached = this.cache.get(cacheKey);
      if (cached) {
        logger.debug(`Using cached embedding for text: ${text.substring(0, 50)}...`);
        return cached.embedding;
      }

      // Clean and prepare text
      const cleanText = this.cleanText(text);
      if (cleanText.length === 0) {
        logger.warn('Empty text provided for embedding generation');
        return new Array(384).fill(0); // Default dimension for nomic-embed-text
      }

      // Generate embedding using Ollama
      const embedding = await this.generateOllamaEmbedding(cleanText, options);
      
      // Cache the result
      const result: EmbeddingResult = {
        embedding,
        dimensions: embedding.length,
        model: options.model || this.defaultModel,
        tokens: this.estimateTokens(cleanText)
      };
      
      this.cache.set(cacheKey, result);
      
      // Clean cache if it gets too large
      if (this.cache.size > 10000) {
        this.clearOldCache();
      }

      return embedding;

    } catch (error) {
      logger.error(`Failed to generate embedding for text: ${text.substring(0, 50)}...`, error);
      // Return zero vector as fallback
      return new Array(384).fill(0);
    }
  }

  async generateBatchEmbeddings(
    texts: string[],
    options: EmbeddingOptions = {}
  ): Promise<number[][]> {
    const embeddings: number[][] = [];
    
    // Process in batches to avoid overwhelming the service
    const batchSize = 10;
    for (let i = 0; i < texts.length; i += batchSize) {
      const batch = texts.slice(i, i + batchSize);
      
      const batchPromises = batch.map(text => this.generateEmbedding(text, options));
      const batchEmbeddings = await Promise.all(batchPromises);
      
      embeddings.push(...batchEmbeddings);
      
      // Small delay between batches to avoid rate limiting
      if (i + batchSize < texts.length) {
        await this.delay(100);
      }
    }

    return embeddings;
  }

  async generateEntityEmbedding(entity: any): Promise<number[]> {
    // Create rich text representation of the entity
    const entityText = this.createEntityText(entity);
    return this.generateEmbedding(entityText);
  }

  async generateRepositoryEmbeddings(repositoryId: string): Promise<void> {
    try {
      logger.info(`Generating embeddings for repository ${repositoryId}`);

      // This would be implemented with database access
      // For now, this is a placeholder that shows the pattern
      
      // 1. Get all entities for the repository
      // 2. Generate embeddings for each entity
      // 3. Store embeddings back to the database
      
      logger.info(`Completed embedding generation for repository ${repositoryId}`);

    } catch (error) {
      logger.error(`Failed to generate repository embeddings:`, error);
      throw error;
    }
  }

  private async generateOllamaEmbedding(
    text: string,
    options: EmbeddingOptions = {}
  ): Promise<number[]> {
    try {
      const response = await axios.post(`${this.ollamaUrl}/api/embeddings`, {
        model: options.model || this.defaultModel,
        prompt: text
      }, {
        timeout: 30000,
        headers: {
          'Content-Type': 'application/json'
        }
      });

      if (response.data && response.data.embedding) {
        let embedding = response.data.embedding;
        
        // Normalize if requested
        if (options.normalize !== false) {
          embedding = this.normalizeVector(embedding);
        }
        
        return embedding;
      } else {
        throw new Error('Invalid response from Ollama embedding service');
      }

    } catch (error) {
      if (error.code === 'ECONNREFUSED') {
        logger.error('Cannot connect to Ollama service. Is it running?');
        throw new Error('Ollama service unavailable');
      }
      
      logger.error(`Ollama embedding request failed:`, error);
      throw error;
    }
  }

  private createEntityText(entity: any): string {
    // Create a rich text representation combining multiple entity fields
    const parts = [];
    
    // Add entity name and type
    parts.push(`${entity.type}: ${entity.name}`);
    
    // Add signature if available
    if (entity.signature) {
      parts.push(`Signature: ${entity.signature}`);
    }
    
    // Add parameters for functions
    if (entity.parameters && entity.parameters.length > 0) {
      parts.push(`Parameters: ${entity.parameters.join(', ')}`);
    }
    
    // Add return type
    if (entity.returnType) {
      parts.push(`Returns: ${entity.returnType}`);
    }
    
    // Add documentation
    if (entity.documentation) {
      parts.push(`Documentation: ${entity.documentation}`);
    }
    
    // Add dependencies
    if (entity.dependencies && entity.dependencies.length > 0) {
      parts.push(`Dependencies: ${entity.dependencies.join(', ')}`);
    }
    
    // Add complexity information
    if (entity.metadata && entity.metadata.complexity) {
      parts.push(`Complexity: ${entity.metadata.complexity}`);
    }

    return parts.join(' | ');
  }

  private cleanText(text: string): string {
    return text
      .replace(/\s+/g, ' ') // Normalize whitespace
      .replace(/[^\w\s\-_.,;:()\[\]{}]/g, '') // Remove special characters
      .trim()
      .substring(0, 8000); // Limit length to avoid token limits
  }

  private estimateTokens(text: string): number {
    // Rough estimation: ~4 characters per token for English text
    return Math.ceil(text.length / 4);
  }

  private normalizeVector(vector: number[]): number[] {
    const magnitude = Math.sqrt(vector.reduce((sum, val) => sum + val * val, 0));
    if (magnitude === 0) return vector;
    return vector.map(val => val / magnitude);
  }

  private getCacheKey(text: string, options: EmbeddingOptions): string {
    const optionsStr = JSON.stringify(options);
    return `${text.substring(0, 100)}_${optionsStr}`;
  }

  private clearOldCache(): void {
    // Simple LRU-like cleanup - remove first 1000 items
    const keys = Array.from(this.cache.keys());
    for (let i = 0; i < 1000 && i < keys.length; i++) {
      this.cache.delete(keys[i]);
    }
  }

  private delay(ms: number): Promise<void> {
    return new Promise(resolve => setTimeout(resolve, ms));
  }

  // Similarity and search methods

  calculateCosineSimilarity(vec1: number[], vec2: number[]): number {
    if (vec1.length !== vec2.length) {
      throw new Error('Vectors must have the same length');
    }

    let dotProduct = 0;
    let norm1 = 0;
    let norm2 = 0;

    for (let i = 0; i < vec1.length; i++) {
      dotProduct += vec1[i] * vec2[i];
      norm1 += vec1[i] * vec1[i];
      norm2 += vec2[i] * vec2[i];
    }

    norm1 = Math.sqrt(norm1);
    norm2 = Math.sqrt(norm2);

    if (norm1 === 0 || norm2 === 0) {
      return 0;
    }

    return dotProduct / (norm1 * norm2);
  }

  async findSimilarEmbeddings(
    queryEmbedding: number[],
    candidateEmbeddings: Array<{ id: string; embedding: number[]; metadata?: any }>,
    topK: number = 10,
    threshold: number = 0.5
  ): Promise<Array<{ id: string; similarity: number; metadata?: any }>> {
    const similarities = candidateEmbeddings.map(candidate => ({
      id: candidate.id,
      similarity: this.calculateCosineSimilarity(queryEmbedding, candidate.embedding),
      metadata: candidate.metadata
    }));

    return similarities
      .filter(item => item.similarity >= threshold)
      .sort((a, b) => b.similarity - a.similarity)
      .slice(0, topK);
  }

  // Health check and diagnostics

  async healthCheck(): Promise<{
    status: 'healthy' | 'unhealthy';
    service: string;
    model: string;
    responseTime: number;
    error?: string;
  }> {
    const startTime = Date.now();
    
    try {
      await this.generateEmbedding('health check test');
      
      return {
        status: 'healthy',
        service: 'ollama',
        model: this.defaultModel,
        responseTime: Date.now() - startTime
      };
      
    } catch (error) {
      return {
        status: 'unhealthy',
        service: 'ollama',
        model: this.defaultModel,
        responseTime: Date.now() - startTime,
        error: error.message
      };
    }
  }

  // Configuration and utility methods

  updateConfiguration(config: {
    ollamaUrl?: string;
    defaultModel?: string;
  }): void {
    if (config.ollamaUrl) {
      this.ollamaUrl = config.ollamaUrl;
    }
    if (config.defaultModel) {
      this.defaultModel = config.defaultModel;
    }
    
    // Clear cache when configuration changes
    this.cache.clear();
  }

  getCacheStats(): {
    size: number;
    maxSize: number;
    hitRate: number;
  } {
    return {
      size: this.cache.size,
      maxSize: 10000,
      hitRate: 0 // Would need to track hits/misses for real implementation
    };
  }

  clearCache(): void {
    this.cache.clear();
  }
}

File: packages/core/src/analyzer-legacy/embedding_service.py
"""
Real Embedding Service for code search and analysis.
Connects to Ollama for generating embeddings.
"""

import logging
import hashlib
import time
from typing import List, Dict, Any, Optional
import requests
from datetime import datetime

logger = logging.getLogger(__name__)

class EmbeddingService:
    """Service for generating embeddings using Ollama."""
    
    def __init__(self, ollama_url: str = "http://localhost:11436", model: str = "nomic-embed-text"):
        """Initialize the embedding service."""
        self.ollama_url = ollama_url.rstrip('/')
        self.model = model
        self.cache = {}  # Simple in-memory cache
        
        # Test connection
        self._test_connection()
    
    def _test_connection(self) -> bool:
        """Test connection to Ollama."""
        try:
            response = requests.get(f"{self.ollama_url}/api/tags", timeout=5)
            if response.status_code == 200:
                models = response.json().get("models", [])
                model_names = [m.get("name") for m in models]
                if self.model in model_names or f"{self.model}:latest" in model_names:
                    logger.info(f"✅ Connected to Ollama. Model {self.model} is available.")
                    return True
                else:
                    logger.warning(f"⚠️ Connected to Ollama, but model {self.model} not found. Available: {model_names}")
                    return False
            else:
                logger.error(f"❌ Ollama responded with status {response.status_code}")
                return False
        except Exception as e:
            logger.error(f"❌ Failed to connect to Ollama at {self.ollama_url}: {e}")
            return False
    
    def get_embedding(self, text: str) -> Optional[List[float]]:
        """Generate embedding for text."""
        if not text.strip():
            return None
            
        # Check cache
        text_hash = hashlib.md5(text.encode()).hexdigest()
        if text_hash in self.cache:
            return self.cache[text_hash]
        
        try:
            # Truncate if too long
            if len(text) > 8000:
                text = text[:8000]
            
            response = requests.post(
                f"{self.ollama_url}/api/embeddings",
                json={"model": self.model, "prompt": text},
                timeout=30
            )
            
            if response.status_code == 200:
                data = response.json()
                embedding = data.get('embedding', [])
                
                if embedding:
                    # Cache the result
                    self.cache[text_hash] = embedding
                    return embedding
                else:
                    logger.error("Empty embedding received from Ollama")
                    return None
            else:
                logger.error(f"Ollama API error: {response.status_code} - {response.text}")
                return None
                
        except Exception as e:
            logger.error(f"Error generating embedding: {e}")
            return None
    
    def get_embeddings_batch(self, texts: List[str]) -> List[Optional[List[float]]]:
        """Generate embeddings for multiple texts."""
        embeddings = []
        for text in texts:
            embedding = self.get_embedding(text)
            embeddings.append(embedding)
            # Small delay to avoid overwhelming the service
            time.sleep(0.1)
        return embeddings
    
    def cosine_similarity(self, vec1: List[float], vec2: List[float]) -> float:
        """Calculate cosine similarity between two vectors."""
        if not vec1 or not vec2 or len(vec1) != len(vec2):
            return 0.0
        
        # Calculate dot product and magnitudes
        dot_product = sum(a * b for a, b in zip(vec1, vec2))
        magnitude1 = sum(a * a for a in vec1) ** 0.5
        magnitude2 = sum(b * b for b in vec2) ** 0.5
        
        if magnitude1 == 0 or magnitude2 == 0:
            return 0.0
        
        return dot_product / (magnitude1 * magnitude2)
    
    def search_similar(self, query_text: str, code_snippets: List[Dict[str, Any]], limit: int = 10) -> List[Dict[str, Any]]:
        """Search for similar code snippets using embeddings."""
        query_embedding = self.get_embedding(query_text)
        if not query_embedding:
            return []
        
        similarities = []
        for snippet in code_snippets:
            snippet_text = snippet.get('content', snippet.get('code', ''))
            snippet_embedding = self.get_embedding(snippet_text)
            
            if snippet_embedding:
                similarity = self.cosine_similarity(query_embedding, snippet_embedding)
                similarities.append({
                    **snippet,
                    'similarity': similarity,
                    'embedding': snippet_embedding
                })
        
        # Sort by similarity (descending) and return top results
        similarities.sort(key=lambda x: x['similarity'], reverse=True)
        return similarities[:limit]
    
    def get_model_info(self) -> Dict[str, Any]:
        """Get information about the current model."""
        try:
            response = requests.get(f"{self.ollama_url}/api/tags", timeout=5)
            if response.status_code == 200:
                models = response.json().get("models", [])
                for model in models:
                    if model.get("name") == self.model or model.get("name") == f"{self.model}:latest":
                        return {
                            "name": model.get("name"),
                            "size": model.get("size", "unknown"),
                            "modified": model.get("modified_at"),
                            "provider": "Ollama",
                            "dimensions": 768,  # nomic-embed-text dimensions
                            "status": "available"
                        }
            
            return {
                "name": self.model,
                "provider": "Ollama",
                "dimensions": 768,
                "status": "unknown"
            }
        except Exception as e:
            logger.error(f"Error getting model info: {e}")
            return {
                "name": self.model,
                "provider": "Ollama",
                "status": "error",
                "error": str(e)
            }
    
    def is_available(self) -> bool:
        """Check if the embedding service is available."""
        return self._test_connection()
    
    def clear_cache(self):
        """Clear the embedding cache."""
        cache_size = len(self.cache)
        self.cache.clear()
        logger.info(f"Cleared embedding cache ({cache_size} entries)")
